tools/certs/Makefile

```diff
@@ -277,3 +279,93 @@ root-key.pem:
 		-extensions req_ext -extfile $(dir $<)/workload.conf \
 		-in $< -out $@
 
+#------------------------------------------------------------------------
+
+.PHONY: %-certs-vm
+
+%-certs-vm: %/vm-workload-cert.pem
+	@make --makefile $(realpath $(firstword $(MAKEFILE_LIST))) clean
+	@echo "done"
+
+%/vm-workload-cert.pem: %/vm-workload.csr  k8s-root-cert.pem k8s-root-key.pem
+	@echo "generating $@"
+	@openssl x509 -req -days $(WORKLOAD_DAYS) \
+		-CA k8s-root-cert.pem  -CAkey k8s-root-key.pem -CAcreateserial\
+		-extensions req_ext -extfile $(dir $<)/vm-workload.conf \
+		-in $< -out $@
+%/vm-workload.csr: L=$(dir $@)
+%/vm-workload.csr: %/key.pem %/vm-workload.conf
+	@echo "generating $@"
+	@openssl req -new -config $(L)/vm-workload.conf -key $< -out $@
+
+%/vm-workload.conf: L=$(dir $@)
+%/vm-workload.conf:
+	@echo "[ req ]" > $@
+	@echo "encrypt_key = no" >> $@
+	@echo "prompt = no" >> $@
+	@echo "utf8 = yes" >> $@
+	@echo "default_md = sha256" >> $@
+	@echo "default_bits = $(INTERMEDIATE_KEYSZ)" >> $@
+	@echo "req_extensions = req_ext" >> $@
+	@echo "x509_extensions = req_ext" >> $@
+	@echo "distinguished_name = req_dn" >> $@
+	@echo "[ req_ext ]" >> $@
+	@echo "subjectKeyIdentifier = hash" >> $@
+	@echo "basicConstraints = critical, CA:false" >> $@
+	@echo "keyUsage = digitalSignature, keyEncipherment" >> $@
+	@echo "extendedKeyUsage = serverAuth, clientAuth" >> $@
+	@echo "subjectAltName=@san" >> $@
+	@echo "[ san ]" >> $@
+	@echo "URI.1 = spiffe://cluster.local/ns/$(L)sa/$(SERVICE_ACCOUNT)" >> $@
+	@echo "[ req_dn ]" >> $@
+	@echo "O = $(INTERMEDIATE_ORG)" >> $@
+	@echo "CN = $(WORKLOAD_CN)" >> $@
+	@echo "L = $(L:/=)" >> $@
+
+#------------------------------------------------------------------------
+
+## <namespace>/token-cert will get a certificate for namespace using the token method. This only requires namespace
+# permissions, will not access istio-system or the certifcate.
+# CA_ADDR can be set to a different location - reflecting the gateway address
+CA_ADDR ?= istiod.istio-system.svc:15012
```

Comment:
is CA_ADDR the key user config that I can leverage in the case I have istiod1 (mesh1) in ns 1, and istiod2 (mesh2) in ns2, both are on c1?

Reply:
CA_ADDR is the 'central istiod address' in your case - the address of Istiod. 

It's the same address that gets configured in the pods, during injection. The steps bellow replicate what the agent is doing.

Reply:
Not sure what you mean about ns1/ns2 - this is the same address that a Pod would use to connect to istiod, wherever it is. 

For example I have a DNS-based cert, and my address is the URL of the gateway where Istiod is exposed ( and I don't need the cert from the config map ).